src/rgw/rgw_lua_utils.cc

```diff
@@ -177,5 +186,34 @@ lua_state_guard::~lua_state_guard() {
   }
 }
 
+void lua_state_guard::runtime_hook(lua_State* L, lua_Debug* ar) {
+  auto now = ceph::real_clock::now();
+  lua_getfield(L, LUA_REGISTRYINDEX, "runtimeguard_max_runtime");
+  auto max_runtime =
+      *static_cast<std::chrono::milliseconds*>(lua_touserdata(L, -1));
+  lua_getfield(L, LUA_REGISTRYINDEX, "runtimeguard_start_time");
+  auto start_time =
+      *static_cast<ceph::real_clock::time_point*>(lua_touserdata(L, -1));
+  lua_pop(L, 2);
+  auto elapsed =
+      std::chrono::duration_cast<std::chrono::milliseconds>(now - start_time);
+
+  if (elapsed > max_runtime) {
+    luaL_error(L, "Lua runtime limit exceeded");
+  }
+}
+
+void lua_state_guard::set_runtime_hook() {
+  lua_pushlightuserdata(state,
+                        const_cast<std::chrono::milliseconds*>(&max_runtime));
+  lua_setfield(state, LUA_REGISTRYINDEX, "runtimeguard_max_runtime");
+  lua_pushlightuserdata(state,
+                        const_cast<ceph::real_clock::time_point*>(&start_time));
+  lua_setfield(state, LUA_REGISTRYINDEX, "runtimeguard_start_time");
+
+  // Check runtime after each line or every 1000 VM instructions
+  lua_sethook(state, runtime_hook, LUA_MASKLINE | LUA_MASKCOUNT, 1000);
```

Comment:
would reccomend disabling the hook in the state dtor: https://github.com/ceph/ceph/pull/61282/files#diff-8bf154c46c53f11c124bff337aef783a6215253ecee05617adc290097fe89ad2L149
(set the mask to zero).
this is to make sure that we don't error during cleanup, which may cause resource leak